src/ert/config/observations.py

```diff
@@ -1,471 +1,193 @@
+            )
+            .set_index(["name", "obs_name", *self._primary_key_values.keys()])
+            .to_xarray()
+        )
+
+    def __len__(self) -> int:
+        return len(self._observations)
+
+
+# Columns used to form a key for observations of the response_type
+ObservationsIndices = {"summary": ["time"], "gen_data": ["index", "report_step"]}
+
+
 class EnkfObs:
     def __init__(self, obs_vectors: Dict[str, ObsVector], obs_time: List[datetime]):
         self.obs_vectors = obs_vectors
         self.obs_time = obs_time
-        self.datasets: Dict[str, xr.Dataset] = {
-            name: obs.to_dataset([]) for name, obs in sorted(self.obs_vectors.items())
+
+        vecs: List[ObsVector] = [*self.obs_vectors.values()]
+
+        gen_obs_ds = _ObsDataset(primary_key=["index", "report_step"])
+        sum_obs_ds = _ObsDataset(primary_key=["time"])
+
+        # Faster to not create a single xr.Dataset per
+        # observation and then merge/concat
+        # this just accumulates 1d vecs before making a dataset
+        for vec in vecs:
+            if vec.observation_type == ResponseTypes.GEN_DATA:
+                for report_step, node in vec.observations.items():
+                    assert isinstance(node, GenObservation)
+                    # wrt typing: Hard to prove to mypy
+                    # that this generic typing works
+                    # ideally this will be abstracted
+                    # a bit differently so typing resolves more nicely
+                    gen_obs_ds.write(
+                        response_name=[vec.response_name] * len(node.values),
+                        obs_name=[vec.observation_name] * len(node.values),
+                        observations=list(node.values),
+                        stds=list(node.stds),
+                        primary_keys={
+                            "report_step": [report_step] * len(node.values),
+                            "index": list(node.indices),
+                        },
+                    )
+
+            elif vec.observation_type == ResponseTypes.SUMMARY:
+                observations = []
+                stds = []
+                dates = []
+                obs_keys = []
+
+                for the_date, obs in vec.observations.items():
+                    assert isinstance(obs, SummaryObservation)
+                    observations.append(obs.value)
+                    stds.append(obs.std)
+                    dates.append(the_date)
+                    obs_keys.append(obs.observation_key)
+
+                sum_obs_ds.write(
+                    response_name=[vec.response_name] * len(observations),
+                    obs_name=obs_keys,
+                    observations=observations,
+                    stds=stds,
+                    primary_keys={"time": dates},
+                )
+            else:
+                raise ValueError("Unknown observation type")
+
+        accumulated_ds: Dict[str, _AccumulatedDataset] = {
```

Comment:
Not sure how much value there is in the protocol, given that everything happens within the scope of this function, but keep it if you feel it adds value.

Reply:
Actually it makes sense to remove it now as the new refactored `_ObsDataset` can be used to type annotate with